### PR TITLE
Triton model versioning

### PR DESCRIPTION
**Pull request by kpedro88**
#### PR description:

This PR is part of #43695. It does two major things:
1. Adds `cmsTritonConfigTool`, a general tool to inspect and edit config.pbtxt files. This tool has some specific modes:
    * `threadcontrol` to handle editing config files for the local fallback server, which was previously done (much less robustly) using `cat`
    * `checksum` to check and inject model checksums for versioning
    * `versioncheck` that runs `checksum` over all `config.pbtxt` files in `$CMSSW_SEARCH_PATH`
2. Updates `TritonClient` to compare model checksums between the local model version (which is known to be the right one) and the remote model version on the server.
3. Adds a unit test to run `cmsTritonConfigTool versioncheck` whenever any cms-data contents are modified.

This PR is kept in draft mode until the other parts of #43695 are complete, otherwise workflows will fail in the meantime.

#### PR validation:

Tested SONIC workflows with this code in place and model checksums in config.pbtxt files, and everything runs correctly. Also tested with a wrong checksum to see that the exception is thrown.

New unit test passes with cms-data update PRs included (see #43695) and fails as expected otherwise.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Not planned to be backported.

### CODE REVIEW

**Comment by kpedro88** [comment]
The `checksum` tool is now merged back into the main `cmsTritonConfigTool`, following the updated plan. Still todo: adding the unit test.

**Comment by kpedro88** [comment]
@smuzaffar in the most recent commit, I've added the global version check as another function of the config tool with a corresponding unit test. Hopefully this is easy to integrate into cms-bot.

**Comment by smuzaffar** [comment]
@kpedro88 , no need to change bot, just add `cmsswdata` as dependency of new unit test i.e.
```
<test name="TestHeterogeneousCoreSonicTritonVersionCheck" command="cmsTritonConfigTool versioncheck">
  <use name="cmsswdata"/>
</test>
```

and when even there is any cms-data external update then bot will also checkout this `HeterogeneousCore/SonicTriton` package and run the run test.

By the way, note that no need to add `${LOCALTOP}/src/HeterogeneousCore/SonicTriton/scripts` , executables from the `scripts` directory are available via PATH.

You can also change `${LOCALTOP}/src/HeterogeneousCore/SonicTriton/test/unittest.sh` to `unittest.sh` in the BuildFile.xml

**Comment by kpedro88** [comment]
@smuzaffar brilliant, thanks! that dramatically simplifies things

**Comment by kpedro88** [comment]
@cms-sw/ml-l2 please take a look, in particular at the "policy": adding a checksum for a new model version is always allowed/required, whereas changing a checksum for an existing model version is noted as not allowed and requiring an exception.

**Comment by valsdav** [comment]
Thanks Kevin, having a look

**Comment by valsdav** [comment]
Hi @kpedro88,  I had a look and I agree with the policy.  Adding the automatic `versioncheck` for cms-data changes is  exactly what we need.   The printout from the tool seems also very clear.   +1 from me!

**Comment by kpedro88** [comment]
all the prerequisites are now merged

**Comment by kpedro88** [comment]
test parameters:
workflows = 10805.31,11624.9001,24834.9001
relvals_opt = --what cleanedupgrade,standard,highstats,pileup,generator,extendedgen,production,identity,ged,machine,premix,nano,gpu,2017,2026

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
@cms-sw/heterogeneous-l2 this is ready for review now

**Comment by kpedro88** [comment]
Comparison differences appear spurious/unrelated, as far as I can tell.

File: HeterogeneousCore/SonicTriton/README.md [hunk 1]
```diff
@@ -148,13 +152,32 @@ Additional details and caveats:
...
+The tool has several modes of operation (each with its own options, which can be viewed using `--help`):
+* `schema`: displays all field names and types for the Triton ModelConfig message class.
+* `view`: displays the field values from a provided `config.pbtxt` file.
+* `edit`: allows changing any field value in a `config.pbtxt` file. Non-primitive types are specified using JSON format.
+* `checksum`: checks and updates checksums for model files (to enforce versioning).
+* `versioncheck`: checks and updates checksums for all `config.pbtxt` files in `$CMSSW_SEARCH_PATH`.
+* `threadcontrol`: adds job- and ML framework-specific thread control settings.
+
+The `edit` mode is intended for generic modifications, and only supports overwriting existing values
+(not modifying, removing, deleting, etc.).
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/README.md:176]: What is the difference between  "overwriting existing values" and "modifying"?

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc [hunk 2]
```diff
@@ -75,22 +82,63 @@ TritonClient::TritonClient(const edm::ParameterSet& params, const std::string& d
...
+  TRITON_THROW_IF_ERROR(client_->ModelConfig(&modelConfigResponse, options_[0].model_name_, options_[0].model_version_),
+                        "TritonClient(): unable to get model config");
+  inference::ModelConfig remoteModelConfig(modelConfigResponse.config());
+
+  std::map<std::string, std::array<std::string, 2>> checksums;
+  size_t fileCounter = 0;
+  for (const auto& modelConfig : {localModelConfig, remoteModelConfig}) {
+    const auto& agents = modelConfig.model_repository_agents().agents();
+    for (const auto& agent : agents) {
+      if (agent.name() == "checksum") {
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/src/TritonClient.cc:121]: I'd find
```cpp
auto agent = std::find_if(agents.begin(), agents.end(), [](auto const& a) { return a.name() == "checksum"; });
if (agents != agents.end()) {
```
a clearer pattern as it describes the intent more clearly (and doesn't add "too much" of boilerplate).

File: HeterogeneousCore/SonicTriton/README.md (hunk 1 above)
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/README.md:176]: If an element is a list: `model.x = [1,2,3]`
then you can overwrite its content, resulting in: `model.x = [4,5,6]`
but you could not perform other operations on the list (insert, delete, append, etc.). You can only provide the final contents of the list. (Trying to support all these operations generically using a command-line interface seemed like overkill.)

Comment (makortel) [reply HeterogeneousCore/SonicTriton/README.md:176]: Ok, makes sense, thanks for the clarification.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by kpedro88** [comment]
please test

**Comment by makortel** [comment]
Comparisons show the usual
* https://github.com/cms-sw/cmssw/issues/43293
* https://github.com/cms-sw/cmssw/issues/39803
* https://github.com/cms-sw/cmssw/issues/43790

**Comment by makortel** [comment]
DAS query error was with 11624.9001, and the workflow is closely related to the content of this PR, so re-test is probably worth it

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by kpedro88** [comment]
@makortel the .9001 workflows appear to have run successfully despite the DAS query issue. The matrix test logs show no issues.

**Comment by makortel** [comment]
Maybe the 11624.9001 (13 TeV TTbar) should be changed to 11634.9001 (14 TeV TTbar) to make the DAS queries to work for that workflow too (in a separate PR)?

**Comment by smuzaffar** [comment]
@kpedro88 , for relvals we run `runTheMatrix.py` without `-i all` that is why wf `11624.9001` [passed](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d132c1/37101/runTheMatrix-results/11624.9001_TTbar_13+2021_SonicTriton/cmdLog) as step1 is running `GEN,SIM`. For Das query , bot runs `runTheMatrix.py -i all`  and for wf `11624.9001`  the required dataset is not available  (`dasgoclient --limit 0 --query 'file dataset=/RelValTTbar_13/CMSSW_13_1_0_pre1-130X_mcRun3_2022_realistic_withNewBSFromEOY2022Data_v2_RV186-v1/GEN-SIM'` does not return any files)

**Comment by makortel** [comment]
Comparisons continue to show the usual
* https://github.com/cms-sw/cmssw/issues/43293
* https://github.com/cms-sw/cmssw/issues/39803
* https://github.com/cms-sw/cmssw/issues/43790

**Comment by kpedro88** [comment]
@makortel :
> Maybe the 11624.9001 (13 TeV TTbar) should be changed to 11634.9001 (14 TeV TTbar) to make the DAS queries to work for that workflow too (in a separate PR)?

That workflow currently does not exist. It will be added in my next PR.

File: HeterogeneousCore/SonicTriton/test/BuildFile.xml [hunk 3]
```diff
@@ -1,5 +1,8 @@
-<test name="TestHeterogeneousCoreSonicTritonProducerCPU" command="${LOCALTOP}/src/HeterogeneousCore/SonicTriton/test/unittest.sh ${LOCALTOP} CPU"/>
-<test name="TestHeterogeneousCoreSonicTritonProducerGPU" command="${LOCALTOP}/src/HeterogeneousCore/SonicTriton/test/unittest.sh ${LOCALTOP} GPU"/>
+<test name="TestHeterogeneousCoreSonicTritonProducerCPU" command="unittest.sh ${LOCALTOP} CPU"/>
+<test name="TestHeterogeneousCoreSonicTritonProducerGPU" command="unittest.sh ${LOCALTOP} GPU"/>
+<test name="TestHeterogeneousCoreSonicTritonVersionCheck" command="cmsTritonConfigTool versioncheck">
+  <use name="cmsswdata"/>
+</test>
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/test/BuildFile.xml:5]: @kpedro88 After this PR has been merged, if it wasn't already in your plans, could you make a test that tries to use an incorrect version/checksum in `cms-data` to make sure this check actually flags that?

Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/test/BuildFile.xml:5]: Yes, this is a good idea.

**Comment by smuzaffar** [comment]
@makortel @kpedro88 , if wf `11624.9001` is going to run for IBs then it will fail due to missing dataset

**Comment by kpedro88** [comment]
@smuzaffar there are no changes to IB workflows in this PR. I just added it here as a custom matrix test in the PR test parameters.

**Comment by makortel** [comment]
> @makortel :
> 
> > Maybe the 11624.9001 (13 TeV TTbar) should be changed to 11634.9001 (14 TeV TTbar) to make the DAS queries to work for that workflow too (in a separate PR)?
> 
> That workflow currently does not exist. It will be added in my next PR.

Right, that's exactly what I was after.

**Comment by rappoccio** [comment]
+1

**Pull request merged**